### ISSUE TITLE
Suggestion: Advanced Trading Feature

### ISSUE THREAD

**Issue by Natesky9** [Railcraft/Railcraft#1093]
Trading with NPCs has always been tedious and not very useful. Most trades are a blatant rip-off, and others are just not useful at all. Villagers just aren't that conscience about the economy. Most mods try to add at least one villager, but the same thing happens to those.

On top of that, there are not many reasons to produce items besides personal use, and not very many times are there a reason to expand transportation to the other areas of the world. In Minecraft, at least.

However, my suggestion is to create a small economy system through the use of worldgen "trading posts" found in villages. These would be some sort of block/structure that would allow the appropriate items to be inserted into, and the payment (some form of currency, most likely emeralds) to be extracted. This would make a special mechanic where the player would have to make special rail lines to continue trading for these resources.

Unlike villager trades, these would be larger orders that would require (and pay out) much more than the standard villager. Another bonus to this system would be that special resources could be traded for through this system as well, allowing for bulk (but not automated) trading through a rail system, and a definite reason to build big

**Comment by CovertJaguar** [comment]
So um... like the Trade Station?

**Comment by Natesky9** [comment]
More in bulk, and with maybe different trades than vanilla. Possibly using just railcraft recipes or vanilla resources for balance reasons? The initial thought was to use this as an incentive to set up rail lines between base and villages, and possibly a way to trade for exclusive/rare items.

Could be a resource sink/something to aim for late game, and a good way to get resources after a ton of infrastructure.

The second thought that came to mind was villager transportation through trains, although that would require more thought to create valid destinations for to-be passengers?

**Comment by dj3520** [comment]
Random idea: Trade Station cart.

**Comment by Forecaster** [comment]
@dj3520 You're gonna have to elaborate on that. No idea what that would do.

**Comment by Natesky9** [comment]
The idea of this was to allow for a sort of NPC economy, one that isn't manual and tedious. It still wouldn't fully automated, as these "contracts" would be completed, and a new one would have to be set up.

As an example of something that could be traded:

One village requests 2000 leather boots for 500 emeralds
Another village is selling 8 leather per emerald
And another village is selling 20 coal per emerald

There is a slight profit if you can efficiently ship all the items to their destination. There is even more if you can supply the items yourself

**Comment by HoldYourWaffle** [comment]
Does this fall inside the scope of railcraft?

**Comment by Natesky9** [comment]
It would be using rails to trade resources. How else/why else would you transport lots of items over long distances?

**Comment by bookerthegeek** [comment]
This honestly could be both inside and outside.

I think you could set it so that a trade station generates in the villager building and is destroyed when broken or picked up.

Then you drive by it on a train and you then drive to another staion in another village. After it returns custom payout. The further away the larger the payout.

If the train is broken or otherwise leaves the trakes the 'contract is broken'

it should be powerful like IC2 crops, but not broken OP like Magical crops.

Configs would be nice for payout to.

**Comment by Natesky9** [comment]
I think it would be simple enough just to make it a trade, instead of having something that complicated.

These trades would be larger versions of the villager trades, almost requiring a train of decent size to complete. The benefit of this over doing a "contract" is that there's no extra data. In this regards, it's nearly the same as a villager trade, with a max use until the trade changes. The trade station will not accept any items that aren't being traded, and the way to detect if a trade is finished, and needs attention, is when the train returns with something other than the payout.

**Comment by CovertJaguar** [comment]
The Trade Stations can trade in bulk. The villagers will continue to trade with it so long as they have a matching trade available and you have provided sufficient stock.

You'll need to be more specific as to why this is insufficient.

**Comment by Natesky9** [comment]
I didn't know the trade station was automatic. I thought it just shortcut you to the villager with that trade.

Anyway, the standard trades are painfully useless as is. There isn't much that can be sold except for crops, and coal, and the only things that are worth buying aren't very useful, except for maybe ender pearls.

The other problem with the current trade station is that it doesn't encourage expansion. All it requires is villagers nearby with the trades you want, and you can just pump emeralds in indefinitely.

My suggestion would make trade stations an exploration goal, and something to expand towards, as well as a unique way to trade for helpful/exclusive resources

**Comment by Forecaster** [comment]
That sounds more like a job for a separate mod to add more villagers with more interesting trades.

**Comment by dj3520** [comment]
@Forecaster basically works like the trade station (so you specify a trade and it will do if there's an equal or better offer) but also interacts with carts in the train for item providing/depositing. The benefit is I can run rains through a village and trade with pretty much everyone in one go.

**Comment by Natesky9** [comment]
Well, the difference between the current trade station, and my suggestion is that the trade station uses the nearby villager trades, whereas I am suggesting a block that has it's own trades. For simplicity, let's say it only has one trade at a time. This trade would have a set payout that is much better than a villager trade, and most likely be requesting some renewable resource in return for emeralds. For every x number of items, it will output y emeralds. A train can then deposit the wares with an unloader, and then retrieve the payout with a loader. There would be an upper limit, where the trade would change, but this limit would be very high, allowing for many transactions.

**Comment by dj3520** [comment]
So you basically want a new block that is a better trade than a normal villager?

**Comment by Natesky9** [comment]
Well, a mechanic that encourages making rails to places on the map to trade resources.

The trade station is just "place near villagers", and can just be placed in base, near some stolen villagers, but the trades are still lame

**Comment by CovertJaguar** [comment]
So how do you stop the whole "place in base" thing?

**Comment by bookerthegeek** [comment]
Make it a world gen only feature, that is removed when the block is placed or broken. 

Botania has a way to do this with their Dayblooms and night shades that spawn in with world generation.

Me speaking Non-Java...
I believe you could make them all a tile entity on world spawn, that dies when you break it and drops a block instead?

**Comment by Natesky9** [comment]
After trying the current trade station in an actual survival world (1.7.10 because no routing yet)
I can say that the trade station is finicky, as in it will look like the trade goes through, but will revert if you try to pick up the payout. Just as well, getting the trades to work is not intuitive.

To avoid the whole "place in base" / "set and forget" piece, There's actually a good solution to this. The trade station can still be crafted, but can not be placed unless it is in a village, maybe even with a maximum number per village, or a range limit between them (maybe one per chunk max)

They could still be generated in villages, maybe even in a worldgen building similar to the train station, or even repurpose it for this exact thing.

The change I would make is to remove the trading hook with villagers, and instead just require some villagers to be nearby. The trades could be supplied by a combination of vanilla resources and addon trades, focusing mostly on renewable resources.

Again, these trades wouldn't be unlimited, and would be completed after several large trains. At that point, they would generate the next trade, and dump the remaining wares into the output

**Comment by HoldYourWaffle** [comment]
You could still just build/emulate a village in your base

**Comment by Natesky9** [comment]
Yes, and that's fine, because the trade there will only be one item being traded. You would still need to network with other trade stations in order to effectively make a profit. 

In fact, it probably is a good idea to set up base near one, and even maybe create a village or two

**Comment by liach** [comment]
Reference #851. These two issues focus on different subjects.
